dask/dataframe/tests/test_groupby.py

```diff
@@ -553,3 +555,57 @@ def test_groupby_multiprocessing():
     with dask.set_options(get=get):
         assert_eq(ddf.groupby('B').apply(lambda x: x),
                   df.groupby('B').apply(lambda x: x))
+
+
+@pytest.mark.parametrize('spec', [
+    {'b': {'c': 'sum'}},
+    {'b': {'c': 'mean'}, 'c': {'a': 'max', 'a': 'min'}},
+    {'b': 'mean', 'c': 'max'},
+    {'b': 'mean', 'c': ['min', 'max']},
+    {'b': np.sum, 'c': ['min', np.max, np.std, np.var]},
+    ['sum', 'mean', 'min', 'max', 'count', 'size', 'std', 'var'],
+    'mean',
+    'var',
+    'std',
+])
+@pytest.mark.parametrize('split_every', [False, 3, None])
+def test_aggregate_examples(spec, split_every):
+    pdf = pd.DataFrame({'a': [1, 2, 6, 4, 4, 6, 4, 3, 7] * 20,
+                        'b': [4, 2, 7, 3, 3, 1, 1, 1, 2] * 20,
+                        'c': [0, 1, 2, 3, 4, 5, 6, 7, 8] * 20},
+                       columns=['c', 'b', 'a'])
+    ddf = dd.from_pandas(pdf, npartitions=20)
+
+    assert_eq(pdf.groupby('a').agg(spec),
+              ddf.groupby('a').agg(spec, split_every=split_every))
+
+    assert_eq(pdf.groupby('a').aggregate(spec),
+              ddf.groupby('a').aggregate(spec, split_every=split_every))
+
+
+def test_aggregate_build_agg_args__reuse_of_intermediates():
+    """Aggregate reuses intermediates. For example, with sum, count, and mean
+    the sums and counts are only calculated once accross the graph and reused to
+    compute the mean.
+    """
+    from dask.dataframe.groupby import _build_agg_args
+
+    no_mean_spec = [
+        ('foo', 'sum', 'input'),
+        ('bar', 'count', 'input'),
+    ]
+
+    with_mean_spec = [
+        ('foo', 'sum', 'input'),
+        ('bar', 'count', 'input'),
+        ('baz', 'mean', 'input'),
+    ]
+
+    no_mean_chunks, no_mean_aggs, no_mean_finalizers = _build_agg_args(no_mean_spec)
+    with_mean_chunks, with_mean_aggs, with_mean_finalizers = _build_agg_args(with_mean_spec)
+
+    assert len(no_mean_chunks) == len(with_mean_chunks)
+    assert len(no_mean_aggs) == len(with_mean_aggs)
+
+    assert len(no_mean_finalizers) == len(no_mean_spec)
+    assert len(with_mean_finalizers) == len(with_mean_spec)
```

Comment:
> Check 3 should always be satisfied. Irrespective of the chosen aggregations only a single dataframe will be generated per partition. Therefore, the number of elements inside the dask graphs should always be same.

Very cool.  Besides being more convenient I anticipate that this will have some very nice performance benefits :)